src/algorithms/calorimetry/CalorimeterClusterRecoCoG.cc

```diff
@@ -338,4 +295,170 @@ std::optional<edm4eic::MutableCluster> CalorimeterClusterRecoCoG::reconstruct(co
   return std::move(cl);
 }
 
+//------------------------------------------------------------------------
+void CalorimeterClusterRecoCoG::associate(
+  const edm4eic::Cluster& cl,
+#if EDM4EIC_VERSION_MAJOR >= 7
+  const edm4eic::MCRecoCalorimeterHitAssociationCollection* mchitassociations,
+#else
+  const edm4hep::SimCalorimeterHitCollection* mchits,
+#endif
+  edm4eic::MCRecoClusterParticleAssociationCollection* assocs
+) const {
+
+  // --------------------------------------------------------------------------
+  // Association Logic
+  // --------------------------------------------------------------------------
+  /*  1. identify all sim hits associated with a given protocluster, and sum
+   *     the energy of the sim hits.
+   *  2. for each sim hit, identify all contributing primary particles, and sum
+   *     their contributed energy.
+   *  3. create an association for each contributiong primary with a weight
+   *     of contributed energy over total energy.
+   */
+
+  // bookkeeping containers
+  std::map<int, std::pair<int, int>> mapMCParToSimIndices;
+  std::map<int, double> mapMCParToContrib;
+  std::vector<int> vecAssocSimHits;
+
+  // --------------------------------------------------------------------------
+  // 1. get associated sim hits and sum energy
+  // --------------------------------------------------------------------------
+  double eSimHitSum = 0.;
+  for (std::size_t iHit = 0; auto clhit : cl.getHits()) {
+
+    // make sure list of associated sim hits is empty
+    vecAssocSimHits.clear();
+
+#if EDM4EIC_VERSION_MAJOR >= 7
+    std::size_t iAssociations = mchitassociations->size();
+    for (std::size_t iAssoc = 0; iAssoc < mchitassociations->size(); ++iAssoc) {
+
+      // if raw hit has same cell ID as reco hit, add sim hit to vector
+      // and increment energy sum
+      if (clhit.getCellID() == (*mchitassociations)[iAssoc].getRawHit().getCellID()) {
```

Comment:
Why do we still need to do matching on CellID? Wasn't the point of the associations that you can now do matching on the actual hit?

Reply:
Ah! That was just for matching rec-to-raw hits. I forgot at the time I could also just match based on ObjectID's (which would probably be preferable):
```
if (clhit.getRawHit().getObjectID() == (*mchitassociations)[iAssoc].getRawHit().getObjectID()) {
```

Reply:
I think it should be just a regular equality even:
```suggestion
      if (clhit == (*mchitassociations)[iAssoc].getRawHit()) {
```
(which is probably exactly the same when compiled, but is neater in code)

Reply:
Ah! I didn't realize you could do that!